### DISCUSSION TITLE
Can VanJS sort without recreating elements?

### DISCUSSION THREAD

**Discussion by lazex** [vanjs-org/van#118] [Q&A] [answered]
I want to sort the elements.
I can't find anything like key prop of React.js in VanJS.

I tried following code.

```js
const { div, button, b, span } = van.tags

const App = () => {
	const items = van.state([{ id: 1, text: "foo" }, { id: 2, text: "bar" }])
	const order = van.state("asc")

	return (
		div(
			button(
				{
					onclick: () => { order.val = order.val === "asc" ? "desc" : "asc" }
				},
				"sort"
			),
			() => {
				const sorted = items.val.toSorted((a, b) => {
					if (order.val === "asc") {
						return a.id - b.id
					} else {
						return b.id - a.id
					}
				})
				return div(
					sorted.map(item => {
						return div({ class: "item" }, b(item.id), span(item.text))
					})
				)
			}
		)
	)
}

van.add(document.body, App())
```

A div element with class `item` will be recreated each time.
If the elements are recreated, some tags like canvas and video cannot maintain state.

Does VanJS support such cases?

**Comment by Tao-VanJS** [comment, marked as answer]
Yes, you can sort child elements without re-rendering the root, but in a different approach. Key prop requires DOM diffing, which would significantly increase the complexity of the framework. Instead, VanJS chose to achieve the result with [stateful binding](https://vanjs.org/tutorial#stateful-binding). With stateful binding, you can customize how your DOM node reacts to state changes. Here are some examples where stateful binding is utilized:

* [Auto-complete](https://github.com/vanjs-org/vanjs-org.github.io/blob/a6d3efaad3e5b1e7634c9a6f87f2a3c0ee220d30/code/auto-complete-stateful-dom-func.ts#L51): suggestion list is changing the selected item without re-rendering itself.
* [VanJS code converter](https://github.com/vanjs-org/vanjs-org.github.io/blob/db9f6cdfc2bcf85cec972f2270870ad670f62917/converter-ui/convert.ts#L69): change the converted code without re-rendering the root DOM (we can't bind the converted result directly to text nodes because it needs to integrate with [Prism.js](https://prismjs.com/) for syntax highlighting).

Thus, you can sort the items without re-rendering the root in a couple lines of custom code with stateful binding, in which way you would have the full transparency and control of how things are rendered.

**Comment by lazex** [comment]
Thanks for your answer.

**Comment by cablehead** [comment]
Thanks for asking this @lazex, I had the exact same question! I'm still skimming vanJS's documentation for the first time, and noticed: [vanX.list](https://github.com/vanjs-org/van/discussions/144) which looks like a prepackaged solution.

**Reply by lazex** [reply]
It looks good.
 Thank you for letting me know.
